src/System.Linq/src/System/Linq/Take.cs

```diff
@@ -21,14 +21,12 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
                 return EmptyPartition<TSource>.Instance;
             }
 
-            IPartition<TSource> partition = source as IPartition<TSource>;
-            if (partition != null)
+            if (source is IPartition<TSource> partition)
```

Comment:
Technically the language would allow this here, but I'm not sure it's more readable.

```csharp
switch (source)
{
case null:
  throw Error.ArgumentNull(nameof(source));

case var _ when (count <= 0):
  return EmptyPartition<TSource>.Instance;

case IPartition<TSource> partition:
  return partition.Take(count);

case IList<TSource> sourceList:
  return new ListPartition<TSource>(sourceList, 0, count - 1);

default:
  return new EnumerablePartition<TSource>(source, 0, count - 1);
}
```

I lean towards staying with `if` statements at least for argument validation.